src/shared/firewall-util-nft.c

```diff
@@ -1130,3 +1175,161 @@ int fw_nftables_add_local_dnat(
         /* table created anew; previous address already gone */
         return fw_nftables_add_local_dnat_internal(ctx->nfnl, add, af, protocol, local_port, remote, remote_port, NULL);
 }
+
+static const char *const nfproto_table[] = {
+        [NFPROTO_ARP]    = "arp",
+        [NFPROTO_BRIDGE] = "bridge",
+        [NFPROTO_INET]   = "inet",
+        [NFPROTO_IPV4]   = "ip",
+        [NFPROTO_IPV6]   = "ip6",
+        [NFPROTO_NETDEV] = "netdev",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(nfproto, int);
+
+void nft_set_context_free_many(NFTSetContexts *s) {
+        assert(s);
+
+        for (size_t i = 0; i < s->n; i++) {
+                free(s->ctx[i].table);
+                free(s->ctx[i].set);
+        }
+
+        s->n = 0;
+        s->ctx = mfree(s->ctx);
+}
+
+int nft_set_context_add(NFTSetContexts *s, int nfproto, const char *table, const char *set) {
+        _cleanup_free_ char *table_dup = NULL, *set_dup = NULL;
+        assert(s);
+        assert(IN_SET(nfproto, NFPROTO_ARP, NFPROTO_BRIDGE, NFPROTO_INET, NFPROTO_IPV4, NFPROTO_IPV6, NFPROTO_NETDEV));
+        assert(table);
+        assert(set);
+
+        table_dup = strdup(table);
+        if (!table_dup)
+                return -ENOMEM;
+
+        set_dup = strdup(set);
+        if (!set_dup)
+                return -ENOMEM;
+
+        NFTSetContext *c;
+        c = greedy_realloc((void **) &s->ctx, s->n + 1, sizeof(NFTSetContext));
```

Comment:
use uppercse GREEDY_REALLOC() macro

Reply:
But it doesn't take array-friendly `n_elem` and `size` as arguments, unlike the lowercase version. There could be `GREEDY_REALLOCARRAY()` with that signature, but it would be more or less equal to
```c
#define GREEDY_REALLOCARRAY greedy_realloc
```